File: backend/app/schemas/rubric.py
"""Rubric Engine Pydantic schemas — model and grading payloads."""

from __future__ import annotations

import uuid

from pydantic import BaseModel, Field


class LevelInput(BaseModel):
    label: str = Field(..., min_length=1, max_length=100)
    description: str | None = None
    points: float = Field(..., ge=0)
    position: int = Field(default=0, ge=0)


class CriterionInput(BaseModel):
    title: str = Field(..., min_length=1, max_length=200)
    description: str | None = None
    weight: float = Field(default=1.0, ge=0)
    position: int = Field(default=0, ge=0)
    levels: list[LevelInput] = Field(default_factory=list)


class RubricCreateRequest(BaseModel):
    title: str = Field(..., min_length=1, max_length=300)
    description: str | None = None
    total_points: int = Field(default=20, ge=0)
    is_template: bool = False
    criteria: list[CriterionInput] = Field(default_factory=list)


class LevelResponse(BaseModel):
    id: str
    criterion_id: str
    label: str
    description: str | None = None
    points: float
    position: int


class CriterionResponse(BaseModel):
    id: str
    rubric_id: str
    title: str
    description: str | None = None
    weight: float
    position: int
    levels: list[LevelResponse] = Field(default_factory=list)


class RubricResponse(BaseModel):
    id: str
    school_id: str
    teacher_id: str
    title: str
    description: str | None = None
    total_points: int
    is_template: bool
    criteria: list[CriterionResponse] = Field(default_factory=list)


class RubricScoreInput(BaseModel):
    criterion_id: uuid.UUID
    level_id: uuid.UUID | None = None
    points_awarded: float = Field(..., ge=0)
    comment: str | None = None


class RubricScoreResponse(BaseModel):
    id: str
    submission_id: str
    criterion_id: str
    level_id: str | None = None
    points_awarded: float
    comment: str | None = None


class RubricResultsResponse(BaseModel):
    submission_id: str
    rubric_id: str | None = None
    total_score: float | None = None
    scores: list[RubricScoreResponse] = Field(default_factory=list)

File: mobile/lib/domain/entities/lms/rubric.dart
class RubricLevel {
  final String id;
  final String label;
  final double score;
  final String description;

  const RubricLevel({
    required this.id,
    required this.label,
    required this.score,
    required this.description,
  });
}

class RubricCriterion {
  final String id;
  final String name;
  final double weight;
  final List<RubricLevel> levels;

  const RubricCriterion({
    required this.id,
    required this.name,
    required this.weight,
    required this.levels,
  });
}

class Rubric {
  final String id;
  final String title;
  final String? description;
  final String? subject;
  final List<RubricCriterion> criteria;
  final double maxScore;
  final String createdBy;
  final String createdAt;
  final String updatedAt;

  const Rubric({
    required this.id,
    required this.title,
    this.description,
    this.subject,
    required this.criteria,
    required this.maxScore,
    required this.createdBy,
    required this.createdAt,
    required this.updatedAt,
  });
}

class RubricGradeEntry {
  final String studentId;
  final String criterionId;
  final String levelId;
  final double score;

  const RubricGradeEntry({
    required this.studentId,
    required this.criterionId,
    required this.levelId,
    required this.score,
  });
}

class RubricGradeResult {
  final String studentId;
  final String rubricId;
  final double totalScore;
  final double maxScore;
  final double percentage;
  final List<RubricGradeEntry> entries;

  const RubricGradeResult({
    required this.studentId,
    required this.rubricId,
    required this.totalScore,
    required this.maxScore,
    required this.percentage,
    required this.entries,
  });
}

class RubricResultsResponse {
  final String rubricId;
  final List<RubricGradeResult> results;

  const RubricResultsResponse({
    required this.rubricId,
    required this.results,
  });
}
